util/include/demangle.hpp

```diff
@@ -0,0 +1,31 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#pragma once
+
+#include <cxxabi.h>
+#include <memory>
+
+struct demangler {
+  static std::string demangle(const char* name) {
+      int status = -4;
+      std::unique_ptr<char, void(*)(void*)> res {abi::__cxa_demangle(name, NULL, NULL, &status), std::free};
+      return (status==0) ? res.get() : name;
```

Comment:
It seems wrong to just accept the mangled name if demangling fails. Shouldn't we throw an exception instead?

Reply:
to tell the truth I'm not sure what's the possible scenario of such a failure when we provide compiler-specified typeid

Reply:
As discussed, we'll put an assert since this should never fail.